src/cript/api/api.py

```diff
@@ -402,6 +406,791 @@ def api_prefix(self):
+                            # # this = {"node":[parent_node0], child_node: [{"uuid": }]
+                            this = {"uuid": v["uuid"]}
+                            API.add_to_dict(removes["payload_json_removes"], child_node, this)
+
+        # ----------//---------
+        # Handle dictionary_item_added
+        for item in data_.get("dictionary_item_added", []):
+            # TODO: come back to this : if its just bracket with nothing else after
+            # if re.match(r"root\['(\w+)'\]", item):
+            #     pass # this doesn't take a uuid
+            pattern = re.compile(r"root\['(\w+)'\]\[(\d+)\]\['uuid'\]")
+            match = pattern.match(item)
+            if match:
+                # if k matches the regex str
+                # if "uuid" in v:  # v itself is a dict so if uuid in v
+                add_uuid = cleaned_modified_[match.group(1)][match.group(2)]["uuid"]
+                added_uuids.append(add_uuid)
+
+            pattern = re.compile(r"root\['(\w+)'\]")
+            match = pattern.match(item)
+            if match:
+                # if k matches the regex str
+                # if "uuid" in v:  # v itself is a dict so if uuid in v
+
+                # add_uuid = cleaned_modified_[match.group(1)][match.group(2)]["uuid"]
+                added_entity = cleaned_modified_[match.group(1)]
+                if isinstance(added_entity, list):
+                    # added_uuids.append(item["uuid"])
+                    for node in cleaned_modified_[match.group(1)]:
+                        added_uuids.append(node["uuid"])
+
+        # Handle dictionary_item_removed
+        for item in data_.get("dictionary_item_removed", []):
+            pattern = re.compile(r"root\['(\w+)'\]\[(\d+)\]\['uuid'\]")
+            match = pattern.match(item)
+            if match:
+                # if k matches the regex str
+                # if "uuid" in v:  # v itself is a dict so if uuid in v
+                rem_uuid = cleaned_modified_[match.group(1)][match.group(2)]["uuid"]
+                removed_uuids.append(rem_uuid)
+                # COME BACK TO WIP HERE
+                child_node = match.group(1)
+                # this = {"node":[parent_node0], child_node: [{"uuid": }]
+                this = {"uuid": rem_uuid}
+                API.add_to_dict(removes["payload_json_removes"], child_node, this)
+
+        # Identify unchanged UUIDs
+        # NOTE: may need to implement # unchanged_uuids = set(added_uuids).intersection(removed_uuids)
+        # right now we need to test if attributes are there
+
+        # print("removed_uuids and added uuids")
+        print("added_uuids")
+        print(added_uuids)
+
+        # removes["payload_json_removes"] = removed_uuids
+        print("removed_uuids")
+        print(removed_uuids)
+
+        return removes
+
+    ###########################################################################################
+    @staticmethod
+    def generate_patch_removes(dict_obj1, obj2):
```

Comment:
I don't think this is what we discussed.

I thought we discussed that we have both old and new project as cript.Node objects.

Plus a map that maps UUID to nodes in the old project tree.

Reply:
so , we have a map of a uuid of a node to the "path" in the json object, (the node with get json) 

essentially , I found the only way to do an "instant access" of a sub object (sub node) by uuid 
is if we take the json format of the first node (where all the uuids will be visible as the node is deserialized)
and then we have a map of that nodes uuid, and the "path" to the sub object or sub node within the json object

because if it is just a node and not an object, you cant create the map this way , 
because youll have 
collection=[<cript.nodes.primary_nodes.collection.Collection object at 0x10690bfe0>], material=[<cript.nodes.primary_nodes.material.Material object at 0x1068fc0e0>]

and really what I need to compare the old and new , for all uuids (or subnodes) ,
 is a uuid map to the node data itself
  (uuid being the key, and value being the path in the json like say a node {"uuid":"123abc"} corresponds to the first material node so in the json object that would be = root["material"][0], and this is what we would compare while "walking" the second node by  the uuids we encounter

Reply:
regarding the find children approach,  this does seem compelling , I thought there wasn't a way to control depth and that it was just all children, but if there's search depth , thats good to know,

though, wouldn't we still need to use deep diff to do comparisson in changes on attributes that only change in value but don't actually have a uuid because if they are not a node and just a string value , for example .

so would any deep diff comparision be within the deep and shallow comparisions?

Reply:
No. I don't think we need deepdiff at all. 
You only need the shallow comparison function that I drafted in the script. 

If there is a shallow difference that node needs a patch. 
And since i am comparing json with depth 1 here, that should include non-node attributes as well

Reply:
Here is my version of a `save` mock-up.
No deep diff needed.

```python
import json
import cript

class LastModifiedDict:
    """
    Just a helper class like a dict, that remembers the order of its entries being modified.
    """
    def __init__(self):
        self._data = {}
        self._order = []

    def __setitem__(self, key, value):
        self._data[key] = value
        if key in self._order:
            self._order.remove(key)
        self._order.append(key)

    def __getitem__(self, key):
        return self._data.get(key)

    def keys_sorted_by_last_modified(self):
        return list(self._order)


# Member function of BaseNode
def shallow_equal(self, other):
    self_child_map = {child.uuid : child for child in self.find_children({}, search_depth=1)}
    del self_child_map[self.uuid]
    other_child_map = {child.uuid : child for child in other.find_children({}, search_depth=1)}
    del other_child_map[other.uuid]

    self_sorted_json = self.get_json(known_uuid=self_child_map.keys(), sort_keys=True, condense_to_uuid={}).json
    other_sorted_json = other.get_json(known_uuid=other_child_map.keys(), sort_keys=True, condense_to_uuid={}).json

    return self_sorted_json == other_sorted_json

# Member function of BaseNode
def deep_equal(self, other):
    self_sorted_json = self.get_expanded_json(sort_keys=True)
    other_sorted_json = other.get_expanded_json(sort_keys=True)
    return self_sorted_json == other_sorted_json


# Member function of API
def save(self, project):
    # In this mockup, we can't actually do the search:
    print("old_node_paginator = self.search(node_type=cript.Project, search_mode=cript.SearchModes.UUID, value_to_search=project)")
    print("old_node_paginator.auto_load_nodes = False")

    try:
        print("old_node_json = next(old_node_paginator)")

        # Just for this mock-up we load from the file instead
        with open("new_project.json") as json_handle:
            old_node_json = json_handle.read()

    except StopIteration: # New Project do PATCH instead
        # Do the PATCH request call.
        return # Return here, since we are done after Posting

    # This is where a patch is needed

    # Load the project in a separate cache
    old_project, old_uuid_map = cript.load_nodes_from_json(nodes_json=old_node_json, _use_uuid_cache={})

    # Check if there is any difference between old and new.
    if deep_equal(project, old_project):
        return # No save necessary, since nothing changed

    delete_uuid = []



    patch_map = LastModifiedDict()

    # Iterate the new project in DFS
    for node in project:
        try:
            old_node = old_uuid_map[node.uuid]
        except KeyError:
            # This node only exists in the new new project,
            # But it has a parent, that will patch it in, so we don't need to do anything
            pass

        # Let's see if we need to delete any children, that existed in the old node, but don't exit in the new node.
        node_child_map = {child.uuid : child for child in node.find_children({}, search_depth=1)}
        old_child_map = {child.uuid : child for child in old_node.find_children({}, search_depth=1)}
        for old_uuid in old_child_map:
            if old_uuid not in node_child_map:
                if old_uuid not in delete_uuid:
                    delete_uuid += [old_uuid]

        # Next we check if the current new node needs a patch
        if not shallow_equal(node, old_node):
            patch_map[node.uuid] = node

    for uuid in reversed(patch_map.keys_sorted_by_last_modified()):
        node = patch_map[uuid]
        print(f"Doing API PATCH for {node.uuid}")

    for uuid in delete_uuid:
        print(f"Doing API Delete for {uuid}")


with cript.API(host="https://lb-stage.mycriptapp.org/") as api:
    with open("new_project.json") as json_handle:
        proj_json = json.load(json_handle)
    proj = cript.load_nodes_from_json(nodes_json=proj_json)
    # Modify deep in the tree
    material_to_modify = proj.collection[0].inventory[0].material[0]
    material_to_modify.name = "sure to be a new name"

    # Delete a node
    proj.material[0].property = []

    save(api, proj)
```

Reply:
You can play with this script.
It should work.

And no deepdiff needed.
Also it looks pretty clean with a design to me.

Reply:
To make this script stand-alone.
A made mock version of the function, they actually need to be member functions.

And also, there are a few commands, that I use to load the JSON from a file instead of actually using a known project from the API.

Reply:
I think the only thing that doesn't work with this is deleting an entire project, but maybe we do a separate function for that anyways.